ibmcloud/TROUBLESHOOTING.md

```diff
@@ -0,0 +1,170 @@
+```
+
+### Create Terraform execution plan
+
+`terraform plan`
+
+```text
+data.ibm_is_subnet.primary: Reading...
+data.ibm_iam_user_policy.user_policies: Reading...
+data.ibm_is_floating_ip.worker: Reading...
+data.ibm_is_instance.worker: Reading...
+data.ibm_iam_user_policy.user_policies: Read complete after 1s [id=david_hay@uk.ibm.com]
+data.ibm_is_floating_ip.worker: Read complete after 3s [id=r022-3db669a0-606e-45e1-b493-084c78bb2714]
+data.ibm_is_subnet.primary: Read complete after 4s [id=02f7-f79cfbb4-6872-4956-8b6a-68f09063833d]
+data.ibm_is_instance.worker: Read complete after 4s [id=02f7_05f17f9c-a1a3-4b92-8f53-3a8c1cec08c6]
+╷
+│ Error: Iteration over null value
+│
+│   on main.tf line 16, in locals:
+│   15:   is_policies_and_roles = flatten([
+│   16:     for policy in data.ibm_iam_user_policy.user_policies.policies: [
+│   17:       for resource in policy.resources: policy.roles
+│   18:       if resource.service == "is" && resource.resource_group_id == "" && resource.resource_instance_id == ""
+│   19:     ]
+│   20:   ])
+│     ├────────────────
+│     │ data.ibm_iam_user_policy.user_policies.policies is null
+│
+│ A null value cannot be used as the collection in a 'for' expression.
+
+For more information, please see the following Slack threads: -
+```
+
+This issue was observed with various releases of the [terraform-provider-ibm](https://github.com/IBM-Cloud/terraform-provider-ibm), including `1.34.0`, `1.41.1` and `1.42.0-beta0`.
+
+The issue was also observed with a very simple Terraform playbook: -
+
+```code
+locals {
+  is_policies_and_roles = flatten([
+    for policy in data.ibm_iam_user_policy.user_policies.policies: [
+    ]
+  ])
+}
+
+data "ibm_iam_user_policy" "user_policies" {
+  ibm_id = var.ibmcloud_user_id
+}
+```
+
+and highlighted the fact that Terraform was unable to retrieve the required User Policies via `data.ibm_iam_user_policy.user_policies.policies` in order to be able to loop across the returned list.
+
+Further investigation suggested that the issue only occurred when the IBM Cloud account had a specific User Policy, as per the following example: -
+
+```text
+Policy ID:   7152fe85-36da-46b5-ab9b-1eac8d212cd5
+Roles:       Reader
+Resources:
+             Service Name   containers-kubernetes
+             Region         us-south
+             namespace      default
+```
+
+If this policy was removed from the account, the plan ran to completion, and the engineer was able to then run `terraform apply`.
+
+If the policy was then added back into the account, the same `A null value cannot be used as the collection in a 'for' expression` exception occurred.
+
+After discussions with the community who maintain the [terraform-provider-ibm](https://github.com/IBM-Cloud/terraform-provider-ibm) project, an issue - [Data source of IAM policies failing to list the policies if the policy has service specific attributes #3801] - was raised, and is being actively worked at time of writing ( May 2022 ).
```

Comment:
The link to issue 3801, which I think should link to here https://github.com/IBM-Cloud/terraform-provider-ibm/issues/3801, fails to format correctly when I open the markdown document for the troubleshooting.

Reply:
Thanks @mattarnoatibm now fixed 🤞

Reply:
Will add an example of using `ibmcloud iam` via a separate commit ...

Reply:
Have added an example using the `ibmcloud` CLI tool